compiler/src/test/java/com/squareup/anvil/compiler/ClassReferenceTest.kt

```diff
@@ -0,0 +1,291 @@
+
+              assertThat(psiFunction.returnTypeOrNull()).isNull()
+              assertThat(descriptorFunction.returnType().fqName).isEqualTo(Unit::class.fqName)
+
+              // That is null as well, because it's not a generic return type and we can't resolve
+              // the return type from a function body. The compiler will take care of this in the
+              // backend.
+              assertThat(psiFunction.resolveGenericReturnTypeOrNull(psiRef)).isNull()
+            }
+            "SomeClass2", "SomeClass3" -> {
+              val psiFunction = psiRef.functions.single()
+              val descriptorFunction = descriptorRef.functions.single { it.name == "hello" }
+
+              assertThat(psiFunction.returnType().fqName).isEqualTo(FqName("kotlin.String"))
+              assertThat(descriptorFunction.returnType().fqName).isEqualTo(FqName("kotlin.String"))
+            }
+            "GenericInterface1" -> {
+              val psiFunction = psiRef.functions.single()
+              val descriptorFunction = descriptorRef.functions.single { it.name == "hello" }
+
+              assertThat(psiFunction.returnTypeOrNull()).isNull()
+              assertThat(descriptorFunction.returnTypeOrNull()).isNull()
+
+              val implementingClass2 = FqName("com.squareup.test.SomeClass2")
+                .toClassReference(psiRef.module)
+                .toPsiReference()
+
+              assertThat(
+                psiFunction.resolveGenericReturnType(implementingClass2).fqName
+              ).isEqualTo(FqName("kotlin.String"))
+
+              assertThat(
+                descriptorFunction.resolveGenericReturnType(implementingClass2).fqName
+              ).isEqualTo(FqName("kotlin.String"))
+
+              // TODO: This is currently failing, but should work in theory. The problem is a mix
+              //  of ClassReferences in the super types being PSI and Descriptor implementations.
+              //  We need to implement resolving types for Descriptors as well. That it's only
+              //  supported for PSI is a smell.
+              // val implementingClass3 = FqName("com.squareup.test.SomeClass3")
+              //   .toClassReference(psiRef.module)
+              //   .toPsiReference()
+              //
+              // assertThat(
+              //   psiFunction.resolveGenericReturnType(implementingClass3).fqName
+              // ).isEqualTo(FqName("kotlin.String"))
+              //
+              // assertThat(
+              //   descriptorFunction.resolveGenericReturnType(implementingClass3).fqName
+              // ).isEqualTo(FqName("kotlin.String"))
+            }
+            "GenericInterface2" -> {
+              assertThat(psiRef.functions).hasSize(0)
+              assertThat(descriptorRef.functions).hasSize(4)
+
+              val descriptorFunction = descriptorRef.functions.single { it.name == "hello" }
+              assertThat(descriptorFunction.returnTypeOrNull()).isNull()
```

Comment:
Is this assert to help verify that the method is using a generic type and make sure that the test source we're compiling doesn't accidentally change to return String directly?

I'm reading it as "function return types for generics are null until they're resolved"

Reply:
It just ensures the expectation that a generic type parameter is null as return type. The return type needs to be resolved for a specific implementation.

Those utilities are a tricky, that's why it also took me so long to refactor them and clean things up. It's confusing :-(